akka-actor-typed/src/main/scala/akka/actor/typed/Extensions.scala

```diff
@@ -144,3 +155,24 @@ trait Extensions {
   def hasExtension(ext: ExtensionId[_ <: Extension]): Boolean
 }
 
+/**
+ * Each extension typically provide a concrete `ExtensionSetup` that can be used in
+ * [[akka.actor.setup.ActorSystemSetup]] when starting the [[ActorSystem]] to replace the default
+ * implementation of the extension. Intended for tests that need to replace
+ * extension with stub/mock implementations.
+ */
+abstract class ExtensionSetup[T <: Extension](
+  val extId:           ExtensionId[T],
+  val createExtension: java.util.function.Function[ActorSystem[_], T])
+  extends Setup
+
+/**
+ * Scala 2.11 API: Each extension typically provide a concrete `ExtensionSetup` that can be used in
+ * [[akka.actor.setup.ActorSystemSetup]] when starting the [[ActorSystem]] to replace the default
+ * implementation of the extension. Intended for tests that need to replace
+ * extension with stub/mock implementations.
+ */
+abstract class AbstractExtensionSetup[T <: Extension](extId: ExtensionId[T], createExtension: ActorSystem[_] ⇒ T)
```

Comment:
My thinking regarding the API here is that this class will not be needed when we remove support for Scala 2.11. Having overloaded function parameters doesn't work well, so the API in `ExtensionSetup` is a plain java.util.function.Function, which can be used from both Java and Scala (>= 2.12).